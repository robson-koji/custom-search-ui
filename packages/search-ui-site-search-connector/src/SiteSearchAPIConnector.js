import { toResultList } from "./ResponseAdapter";
import { adaptFacetConfig, adaptFilterConfig } from "./RequestAdapter";

function _request(documentType, engineKey, method, path, params) {
  const headers = new Headers({
    "Content-Type": "application/json"
  });

  return fetch(`https://search-api.swiftype.com/api/v1/public/${path}`, {
    method,
    headers,
    body: JSON.stringify({
      engine_key: engineKey,
      ...params
    }),
    credentials: "include"
  }).then(response => {
    return response.json().then(json => {
      return toResultList(json, documentType);
    });
  });
}

export default class SiteSearchAPIConnector {
  constructor({ documentType, engineKey }) {
    this.documentType = documentType;
    this.engineKey = engineKey;
    this._request = _request.bind(this, documentType, engineKey);
  }

  click() {
    // TODO
  }

  search(searchTerm, searchOptions) {
    const safeDestructureSort = obj => Object.entries(obj || {})[0] || [];

    const { facets, filters, page, sort, ...rest } = searchOptions;
    const [sortField, sortDirection] = safeDestructureSort(sort);
    const updatedFacets = adaptFacetConfig(facets);
    const updatedFilters = adaptFilterConfig(filters);
    return this._request("POST", "engines/search.json", {
      ...rest,
      ...(page &&
        page.size && {
          per_page: page.size
        }),
      ...(page &&
        page.current && {
          page: page.current
        }),
      ...(sortDirection && {
        sort_direction: {
          [this.documentType]: sortDirection
        }
      }),
      ...(sortField && {
        sort_field: {
          [this.documentType]: sortField
        }
      }),
      ...(updatedFacets && {
        facets: {
          [this.documentType]: updatedFacets
        }
      }),
      ...(updatedFilters && {
        filters: {
          [this.documentType]: updatedFilters
        }
      }),
      q: searchTerm
    });
  }
}
