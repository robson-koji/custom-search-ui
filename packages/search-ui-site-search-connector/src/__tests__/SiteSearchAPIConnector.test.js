import SiteSearchAPIConnector from "..";

import exampleAPIResponse from "../../resources/example-response.json";

function fetchResponse(response) {
  return Promise.resolve({
    json: () => Promise.resolve(response)
  });
}

beforeEach(() => {
  global.Headers = jest.fn();
  global.fetch = jest.fn().mockReturnValue(fetchResponse(exampleAPIResponse));
});

const engineKey = 12345;
const documentType = "national-parks";
const params = {
  documentType,
  engineKey
};

it("can be initialized", () => {
  const connector = new SiteSearchAPIConnector(params);
  expect(connector).toBeInstanceOf(SiteSearchAPIConnector);
});

describe("#search", () => {
  function subject(searchTerm = "searchTerm", options = {}) {
    const connector = new SiteSearchAPIConnector(params);
    return connector.search(searchTerm, options);
  }

  it("will call the API with the correct body params", async () => {
    await subject("searchTerm", {
      arbitraryParameter: "shouldRemain",
      page: {
        current: 1,
        size: 10
      },
      sort: {
        name: "desc"
      },
      facets: {
        states: {
          type: "value",
          size: 30
        }
      },
      filters: {
        all: [
          {
            title: ["Acadia"]
          },
          {
            title: ["Grand Canyon"]
          },
          {
            world_heritage_site: ["true"]
          }
        ]
      }
    });
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(JSON.parse(global.fetch.mock.calls[0][1].body)).toEqual({
      engine_key: engineKey,
      arbitraryParameter: "shouldRemain",
      page: 1,
      per_page: 10,
      filters: {
        [documentType]: {
          title: {
            type: "and",
            values: ["Acadia", "Grand Canyon"]
          },
          world_heritage_site: {
            type: "and",
            values: ["true"]
          }
        }
      },
      sort_direction: {
        "national-parks": "desc"
      },
      sort_field: {
        "national-parks": "name"
      },
      facets: {
        "national-parks": ["states"]
      },
      q: "searchTerm"
    });
  });

  it("will correctly format an API response", async () => {
    const response = await subject();
    expect(response).toMatchSnapshot();
  });
});

describe("#click", () => {
  function subject(clickData) {
    const connector = new SiteSearchAPIConnector(params);
    return connector.click(clickData);
  }

  it("will call the API with the correct body params", async () => {
    const query = "test";
    const documentId = "12345";

    await subject({
      query,
      documentId
    });

    expect(global.fetch).toHaveBeenCalledTimes(1);
    const url = global.fetch.mock.calls[0][0];
    const urlWithoutTimestamp = url.replace(/&t=\d*/, "").replace(/t=\d*&/, "");
    expect(urlWithoutTimestamp).toEqual(
      `https://search-api.swiftype.com/api/v1/public/analytics/pc?engine_key=${engineKey}&q=${query}&doc_id=${documentId}`
    );
  });
});
