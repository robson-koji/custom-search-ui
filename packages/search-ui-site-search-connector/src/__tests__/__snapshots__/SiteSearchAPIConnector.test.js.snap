// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#search will correctly format an API response 1`] = `
ResultList {
  "info": Object {
    "facets": Object {
      "states": Array [
        Object {
          "data": Array [
            Object {
              "count": 2,
              "value": "Maine",
            },
          ],
          "type": "value",
        },
      ],
    },
    "meta": Object {
      "page": Object {
        "current": 1,
        "size": 20,
        "total_pages": 1,
        "total_results": 2,
      },
      "request_id": "",
      "warnings": Array [],
    },
  },
  "rawInfo": Object {
    "facets": Object {
      "states": Array [
        Object {
          "data": Array [
            Object {
              "count": 2,
              "value": "Maine",
            },
          ],
          "type": "value",
        },
      ],
    },
    "meta": Object {
      "page": Object {
        "current": 1,
        "size": 20,
        "total_pages": 1,
        "total_results": 2,
      },
      "request_id": "",
      "warnings": Array [],
    },
  },
  "rawResults": Array [
    Object {
      "acres": Object {
        "raw": 49057.36,
      },
      "external_id": Object {
        "raw": "9000",
      },
      "id": Object {
        "raw": "5beb474f8db2315427beecc7",
      },
      "nps_link": Object {
        "raw": "https://www.nps.gov/acad/index.htm",
      },
      "states": Object {
        "raw": Array [
          "Maine",
        ],
      },
      "title": Object {
        "raw": "Acadia",
        "snippet": "<em>Acadia</em>",
      },
      "updated_at": Object {
        "raw": "2018-11-13T21:51:11+00:00",
      },
    },
    Object {
      "acres": Object {
        "raw": 49057.36,
      },
      "external_id": Object {
        "raw": "0",
      },
      "id": Object {
        "raw": "5beb47b8a1f2532eb09be102",
      },
      "nps_link": Object {
        "raw": "https://www.nps.gov/acad/index.htm",
      },
      "states": Object {
        "raw": Array [
          "Maine",
        ],
      },
      "title": Object {
        "raw": "Acadia",
        "snippet": "<em>Acadia</em>",
      },
      "updated_at": Object {
        "raw": "2018-11-13T21:52:56+00:00",
      },
    },
  ],
  "results": Array [
    ResultItem {
      "data": Object {
        "acres": Object {
          "raw": 49057.36,
        },
        "external_id": Object {
          "raw": "9000",
        },
        "id": Object {
          "raw": "5beb474f8db2315427beecc7",
        },
        "nps_link": Object {
          "raw": "https://www.nps.gov/acad/index.htm",
        },
        "states": Object {
          "raw": Array [
            "Maine",
          ],
        },
        "title": Object {
          "raw": "Acadia",
          "snippet": "<em>Acadia</em>",
        },
        "updated_at": Object {
          "raw": "2018-11-13T21:51:11+00:00",
        },
      },
    },
    ResultItem {
      "data": Object {
        "acres": Object {
          "raw": 49057.36,
        },
        "external_id": Object {
          "raw": "0",
        },
        "id": Object {
          "raw": "5beb47b8a1f2532eb09be102",
        },
        "nps_link": Object {
          "raw": "https://www.nps.gov/acad/index.htm",
        },
        "states": Object {
          "raw": Array [
            "Maine",
          ],
        },
        "title": Object {
          "raw": "Acadia",
          "snippet": "<em>Acadia</em>",
        },
        "updated_at": Object {
          "raw": "2018-11-13T21:52:56+00:00",
        },
      },
    },
  ],
}
`;
